nubis/bin/consul2unixusers.py

```diff
@@ -0,0 +1,161 @@
+    consul_port = _config.get('consul', 'port')
+    consul_scheme = _config.get('consul', 'scheme')
+    consul_path = _config.get('consul', 'path')
+    c = consul.Consul(host=consul_host, port=consul_port, scheme=consul_scheme)
+    x = c.kv.get(consul_path, keys=True, separator='/')
+    if x[1] == None:
+        print "ERROR: No keys returned from Consul at %s://%s:%s/%s" % (
+            consul_scheme, consul_host, consul_port, consul_path)
+        sys.exit(-1)
+    ldapCNs = x[1]
+    for key in ldapCNs:
+        uid = c.kv.get(key + 'uid')[1]["Value"]
+        users[uid] = {}
+        users[uid]["homeDirectory"] = c.kv.get(key + 'homeDirectory')[1]["Value"]
+        users[uid]["loginShell"] = c.kv.get(key + 'loginShell')[1]["Value"]
+        users[uid]["mail"] = c.kv.get(key + 'mail')[1]["Value"]
+        users[uid]["uidNumber"] = c.kv.get(key + 'uidNumber')[1]["Value"]
+        users[uid]["sshPublicKey"] = {}
+        sshkeys = c.kv.get(key + 'sshPublicKey/', keys=True, separator='/')[1]
+        for keypath in sshkeys:
+            keynumber = keypath.split('/')[-1]
+            users[uid]["sshPublicKey"][keynumber] = c.kv.get(keypath)
+    return users
+
+def user_exists(user):
+    try:
+        pwd.getpwnam(user)
+    except KeyError:
+        return False
+    return True
+
+def adduser(username, userdata):
+    cmd = "%s -d '%s' -s '%s' -u '%s' -G sudo -m -U %s" % (
+        USERADD, userdata["homeDirectory"], userdata["loginShell"],
+        userdata["uidNumber"], username
+    )
+    if DRYRUN:
+        print cmd
+        return True
+    else:
+        return os.system(cmd)
+
+def writeSSHKeysForUser(username, userdata):
+    # Does their home directory exist? If not, abort, as it should have been
+    # created by adduser()
+    homedir = userdata["homeDirectory"]
+    if not os.path.isdir(homedir):
+        print "ssh key writeout aborted for Unix user '%s': homedir '%s' doesn't exist." % \
+        (username, homedir)
+        return False
+
+    # Does their ~/.ssh directory exist? If not, create it"
+    sshdir = homedir + "/.ssh"
+    if not os.path.isdir(sshdir):
+        mkdircmd = "%s %s" % (MKDIR, sshdir)
+        chowncmd = "%s %s.%s %s" % (CHOWN, username, username, sshdir)
+        chmodcmd = "%s %s %s" % (CHMOD, 700, sshdir)
+        if DRYRUN:
+            print mkdircmd
+            print chowncmd
+            print chmodcmd
+        else:
+            os.system(mkdircmd)
+            os.system(chowncmd)
+            os.system(chmodcmd)
+
+    # Do they already have a ~/.ssh/authorized_keys file? If not, create it
+    authfilepath = sshdir + "/authorized_keys"
+    if not os.path.isfile(authfilepath):
+        authchowncmd = "%s %s.%s %s" % (CHOWN, username, username, authfilepath)
+        if DRYRUN:
+            print "os.mknod('%s', 0600)" % authfilepath
+            print authchowncmd
+        else:
+            os.mknod(authfilepath, 0600)
+            os.system(authchowncmd)
+
```

Comment:
I think this is in scope. If we are deploying this tool that presupposes that we are controlling these files. I would consider it an all or nothing approach.

Now I would agree that creating a hybrid or multi-tool approach is out of scope. It is this limited and rigid application of scope that IMO makes ensuring permissions, removing stale keys, removing users, etc.. in scope for this tool.